### PR TITLE
[installinator] implement progress reports, client side

### PR DESCRIPTION
**Pull request by sunshowers**
Implement client-side progress reports in the installinator.

Introduce a new crate called `installinator-common`, with a `ProgressReport` type that represents a set of events. A `ProgressReport` consists of:
* completion events, which are important events that indicate completion
* progress events, which indicate general progress (for now, we only send the last event, though the data model supports multiple events)

Each event, and each report, has a timestamp associated with it (as a duration since the start of the installinator process). This helps potential users get a good sense of when things happened.

The `ProgressReport` type is much easier to work with than the types generated by progenitor (e.g. real `Duration`s), so use them in both the client and server and convert over while serializing the request in the client.

TODO:

* [x] Extend property-based tests.
  * I think we can write more extensive tests, but the current tests ensure that we receive success messages at least, and we can come back to this after PVT1.
* [ ] Progress and completion reports for writes (this is deferred to a future PR that implements server-side support)
* [ ] Test end-to-end with real wicketd (this is deferred to a future PR that implements server-side support, on top of #2411)
* [ ] Make a single task that discovers peers, share it across both artifact fetches and progress reports (deferred to post-PVT1)

Questions:

* [ ] We currently reach out to every peer rather than existing once we see the first one that accepts our request. Is that correct? (This can be changed after the PVT1 deadline.)
* [ ] Is it really worth exposing `ProgressReport` in the JSON schema? What if instead of that we just use `UntypedBody` then serialize/deserialize to `installinator_common::ProgressReport` on the client/server respectively? I guess since this will require cross-version compatibilty (new installinator talking to old wicketd), having a defined, inspectable schema makes it easier to ensure that it won't break over time.
* [ ] Is it worth adding some jitter to event sends?

### CODE REVIEW

File: wicketd/src/artifacts.rs [hunk 1]
```diff
@@ -132,6 +134,14 @@ impl ArtifactGetter for WicketdArtifactStore {
             buf_list.into_iter().map(|bytes| Ok::<_, Infallible>(bytes)),
         )))
     }
+
+    async fn report_progress(
+        &self,
+        _update_id: Uuid,
+        _report: ProgressReport,
+    ) -> Result<ProgressReportStatus, HttpError> {
+        todo!("implement server-side support for events")
```
Comment (sunshowers) [inline wicketd/src/artifacts.rs:143]: Working on this next.

File: installinator/src/reporter.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            }
+        };
+
+        // We could use StreamExt::any() here, but we're choosing to avoid
+        // short-circuiting for now in favor of reaching out to all peers.
+        //
+        // TODO: is this correct? If a peer takes too long this might end up
+        // stalling updates for everyone -- feels wrong. The tradeoff is that if
+        // two servers both say, only one of them will deterministically get the
+        // update. Need to decide post-PVT1.
```
Comment (jgallagher) [inline installinator/src/reporter.rs:148]: Two thoughts:

1. I think this reaffirms my belief that `broadcast_report` should return a `FuturesUnordered` so we can send all the reports concurrently - I think this `collect()` is effectively building one by hand, although I'm not sure whether the requests are sent sequentially with collect?
2. My gut feeling is waiting for the first successful response seems right: the only peer that should know our update ID is the one that generated it, so we shouldn't see multiple successful responses. "Shouldn't" is carrying a lot of weight there (bugs, malicious peer, etc.) though, I'll admit!

**Review by jgallagher — APPROVED** [review]
This looks great!

File: installinator/src/peers.rs [hunk 3]
```diff
@@ -304,18 +382,60 @@ impl Peers {
             }
         }
     }
+
+    pub(crate) fn broadcast_report(
+        &self,
+        update_id: Uuid,
+        report: ProgressReport,
+    ) -> impl Stream<Item = Result<(), ClientError>> + Send + '_ {
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply installinator/src/peers.rs:390]: The buffer_unordered call inside this function implements concurrency already (it internally uses FuturesUnordered). buffer_unordered also lets you set a limit on the amount of concurrency (in this case 8).

File: installinator/src/reporter.rs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+            let mut events_done = false;
+
+            loop {
+                tokio::select! {
+                    _ = interval.tick() => {
+                        self.on_tick().await;
+                    }
+
+                    event = self.event_receiver.recv(), if !events_done => {
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply installinator/src/reporter.rs:80]: If I'm understanding you correctly, that's how the select macro already works: it first checks if events_done is true, and only polls the future if it isn't.

File: installinator/src/reporter.rs (hunk 2 above)
Comment (sunshowers) [reply installinator/src/reporter.rs:148]: 1. Addressed above: it's already concurrent thanks to buffer_unordered.
2. Yeah, bugs and malicious peers are my main concern here. (We don't have to think about this right now, though—we can always revisit this decision in the future)

File: installinator/src/peers.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgallagher) [reply installinator/src/peers.rs:390]: Ah, I missed that! Thanks.

File: installinator/src/reporter.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgallagher) [reply installinator/src/reporter.rs:80]: Huh, TIL. I always assumed it worked like `match` ifs and was evaluated after the arm already matched.

File: installinator/src/peers.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply installinator/src/peers.rs:390]: TIL! That's awesome.

**Review by andrewjstone — COMMENTED** [review]
🚀

**Pull request merged**